text/0002-new-architecture.md

```diff
@@ -0,0 +1,243 @@
+* Start Date: 2022-07-21
+* RFC Type: informational
+* RFC PR: https://github.com/getsentry/rfcs/pull/2
+
+# Summary
+
+This document is a living piece that is updated with some collected thoughts on what likely projects
+are over a multi year horizon to improve Sentry's underpinnings.
+
+# Motivation
+
+We are running into scaling limitations on our current infrastructure and as such some larger
+decisions have to be made about the future of our codebase.  As we are talking about creating
+more Sentry deployments in different regions and for smaller clusters, some of these problems
+are becoming more pressing as they create complexity for each added installation.
+
+# Expected Workstreams
+
+This captures all expected streams of work that are in the scope of overhauling Sentry's
+architecture.
+
+## Processing Pipeline
+
+These are long running concerns with the processing pipeline that should be addressed.
+
+### Isolate Processing Pipeline out of Monolith
+
+The celery tasks, event manager and related functionality used by the processing pipeline is living
+alongside the rest of the Sentry code in the `sentry` monolith.  This means that it's quite chaotic
+and many unintended cross dependencies are regularly introduced into the code base.  As some of the
+processing pipeline wants to move out of the monolith for better scalability (in some extreme cases
+even move out of Python as host language) it would be helpful by moving all of the processing pipeline
+into a `sentry_pipeline` package.  It would be permissible for `sentry_pipeline` to import `sentry`
+but not the other way round.
+
+This will cause issues in the testsuite as the sentry test suite currently tries to create events
+regularly and pragmatic solutions for this need to be found.
+
+### Remove Pickle in Rabbit
+
+We still use pickle as serialization format for Celery which means that it's not possible for code
+outside of the Sentry monolith to dispatch tasks.  This is both an issue for the separation of
+`sentry` and `sentry_pipeline` (as import paths are changing) as well as it causes problems for
```

Comment:
While I absolutely support moving away from pickle, I would challenge that this is a blocker for renaming `sentry` to `sentry_pipeline`. Celery tasks are explicitly named using the `instrumented_task` decorator, and while the names usually correspond to teh import path, you can move tasks around and rename modules without breaking task producers.

The only place where I think actual import paths are serialized for Celery today is when a complex object is pickled as a task argument. There is prior art to restricting what objects may be pickled in `load-it-up.py`. A custom pickle serializer wired up to Celery that warns if a non-JSON-serializable object is being serialized should be step 0 before moving to a different serialization format. Ideally we'd be doing the same for org/project/global options.

This step 0 will require a lot of coordination and followup tasks for many product areas as I am fairly sure we pickle complex objects in task arguments.

Reply:
The issue is not the naming of the task but the arguments. If I today want to dispatch a task containing a model from outside the sentry codebase, I need to create a message that pickles into a django model from within the sentry codebase. The only realistic and portable way to do this is to import the monolith, create the instance and dispatch the task that way.

Reply:
The non JSON safe detection code BTW is already in Sentry. The issue is that the tasks are still passing these complex objects in practice. That's a thing we need to start addressing before we can even entertain the idea of moving off it.